### PR TITLE
`@otherwise` of `@when`

### LINKED ISSUE #40: else branch of @when
Currently `@when` can only handle with a single branch with multiple destructuring, but can do nothing to the scenarios an `else` is needed. It turns out `@when` is not an equivalent of if-let in Rust.

The enhancement should be performed as soon as possible, but the obstacle is the syntax design.
I don't have an idea about how to make an elegant and concise one as if-let equivalent.

Any suggestions are welcome.

### PR DESCRIPTION
**Pull request by inkydragon**
try to fix #40 

~~Syntax design need more discuzz.~~

Need more test.

This is my proposal: [patch for `MLStyle.jl`](https://gist.github.com/inkydragon/0787b44eb41461c3e6c5898f6042a2dd#file-testset-jl)

---

pr state: **POC**(proof of concept)

### CODE REVIEW

**Comment by thautwarm** [comment]
Cool boy, you've done a lot about this.
I have some issues with this: https://github.com/thautwarm/MLStyle.jl/pull/56/files#diff-d4c083746da5bfbc6b58f97dfca75141R104

Why this
```julia
     @when (a, 1) = x begin
                a
      @when (b, 2)
                (:b, b)
      @when (c, 3)
                (:c, c)
      end
```
instead of
```julia
     @when (a, 1) = x begin
                a
      @when (b, 2) = x
                (:b, b) = x 
      @when (c, 3) = x
                (:c, c)
      end
```

It seems that your implementation of `@when` is covered by `@match`.

The reason why we  need `@when` other than solely `@match` is, a value to match always corresponds to many patterns/cases but there're some use cases that multiple values to match correspond to only one/multiple patterns.

**Comment by thautwarm** [comment]
I think your coding style is pretty attractive. I'll write some about coding style with MLStyle and recommend yours.

**Comment by thautwarm** [comment]
https://github.com/thautwarm/MLStyle.jl/pull/56/files#diff-8065cdff82bfb8220304990b844689c9R121

LineNumberNode shouldn't be ignored, it's crucial for error reporting.

File: src/MLStyle.jl [hunk 1]
```diff
@@ -94,13 +94,108 @@ p                end
...
 You should pass an `Expr(:let, ...)` as the first argument.
 """
 function gen_when(let_expr, source :: LineNumberNode, mod :: Module)
     @match let_expr begin
-        Expr(:let, Expr(:block, bindings...) ||  a && Do(bindings = [a]), ret) =>
+        # `@otherwise`
+        Expr(:let, 
+            Expr(:block, bindings...) || a && Do(bindings = [a]),
+            Expr(:block, 
+                ::LineNumberNode,
```
Comment (thautwarm) [inline src/MLStyle.jl]: Also, we shouldn't expect there is a LineNumberNode, for even if there's no LineNumberNode we can still compile the block expression.

File: src/MLStyle.jl [hunk 2]
```diff
@@ -94,13 +94,108 @@ p                end
...
+        _ => nothing
+    end
+end
+
+function RecursionMatch(block, bds=[], rets=[], default=nothing)
+    @match block begin
+        # empty block
+        Expr(:block) => (bds, rets, default)
+        
+        # skip `LineNumberNode`
```
Comment (thautwarm) [inline src/MLStyle.jl]: LineNumberNode shouldn't be ignored, it's crucial for error reporting.

File: src/MLStyle.jl [hunk 3]
```diff
@@ -94,13 +94,108 @@ p                end
...
+        Expr(:block, 
+            ::LineNumberNode,
+            remain...
+        ) => 
+            RecursionMatch(Expr(:block, remain...), bds, rets, default)
+        
+        # match `@when`s at middle
+        Expr(:block, 
+            MacroSplit{"when"}(bd),
+            ::LineNumberNode,
```
Comment (thautwarm) [inline src/MLStyle.jl]: Don't rely on if there's a LineNumberNode, some packages like Test.jl will remove LineNumberNode in Expr except those of `macrocall`s.

File: src/MLStyle.jl [hunk 4]
```diff
@@ -94,13 +94,108 @@ p                end
...
+        # match `@otherwise` at end
+        Expr(:block, 
+            IsMacro{"otherwise"}(),
+            ::LineNumberNode,
+            default
+        ) => (bds, rets, default)
+
+        # error handle
+        
+        _ => @syntax_err "Expect a :block."
```
Comment (thautwarm) [inline src/MLStyle.jl]: Yours seem to not support this:

```julia
@when let a = b
      stmt1
      stmt2
@when  c = d
       stmt3
       stmt4
end
```

**Comment by thautwarm** [comment]
Pull the branch `https://github.com/thautwarm/MLStyle.jl/tree/when-else` and check what I've done for you.

**Comment by thautwarm** [comment]
You've done a great job, but not all parts make sense.

File: src/MLStyle.jl (hunk 4 above)
Comment (inkydragon) [reply src/MLStyle.jl]: Done.

**Comment by thautwarm** [comment]
In fact there's still some limitation, but I think it sufficient in current stage. I'll make some tiny modifications later.

**Pull request merged**